zerver/views/streams.py

```diff
@@ -607,16 +612,41 @@ def get_streams_backend(
     return json_success({"streams": streams})
 
 @has_request_variables
-def get_topics_backend(request: HttpRequest, user_profile: UserProfile,
+def get_topics_backend(request: HttpRequest, maybe_user_profile: Union[UserProfile, AnonymousUser],
                        stream_id: int=REQ(converter=to_non_negative_int,
-                                          path_only=True)) -> HttpResponse:
-    (stream, recipient, sub) = access_stream_by_id(user_profile, stream_id)
+                                          path_only=True),
+                       narrow: Optional[List[Dict[str, Any]]]=REQ('narrow', converter=narrow_parameter, default=None),
+) -> HttpResponse:
+    if is_web_public_narrow(narrow) and not maybe_user_profile.is_authenticated:
+        is_web_public_query = True
+        user_profile: Optional[UserProfile] = None
+    else:
+        is_web_public_query = False
+        assert isinstance(maybe_user_profile, UserProfile)
+        user_profile = maybe_user_profile
+        assert user_profile is not None
+
+    if is_web_public_query:
+        realm = get_realm_from_request(request)
+        if realm is None:
+            return json_error(_("Invalid subdomain."))
+
+        stream = get_stream_by_id_in_realm(stream_id, realm)
+        if not stream.is_web_public:
+            return json_error(_("Invalid stream"))
```

Comment:
Let's move this block into a `access_web_public_stream(stream_id, realm)` raising JsonableError; we don't want any direct access like this in `streams.py`.